### ISSUE TITLE
[ENH] "condition" on time series object for use in conditional transform, forecast, or fallback

### ISSUE THREAD

**Issue by fkiraly** [sktime/sktime#5936]
There are multiple cases in the codebase where it would be neat if a user could specify a condition easily, such as "the time series is stationary according to test X", or "the time series contains NA", or "the time series is unstable for specific meaning Y".

Examples:
* `TransformIf` transformer, which transforms if a condition is met, e.g., differences if non-stationary.
* `FallbackForecaster`, hypothetical extension from discussion in this PR https://github.com/sktime/sktime/pull/5924 - where we consider nans or fast divergence as failure, not just exceptions

The "natural" or at least closest scitype in `sktime` currently is that of the paramete restimator, scitype code `param_est`.

Besides adding `TransformIf`-like arguments in every such forecaster, there are further options which might lead to neater syntax:

* a `Condition` parameter estimator, which takes the parts from `TransformIf` that define the condition
* a new `Condition` object, not a parameter estimator
* a regex string that can be combined with a parameter estimator, both args must be passed and form the "condition" together

maybe there are also other good options here? FYI @ninedigits, @yarnabrina

**Comment by ninedigits** [comment]
I'm wondering if such a change would be best implemented within the BaseForecaster? Haven't fully thought it through. But you could add in additional validation steps at the fit and predict stages that default to do nothing, leaving it up to the various different inheriting estimators to define what they are. 

```python3
def _validate_y_out(self, y_out):
    """Validation step to be overwritten by user or estimator""" 
    return y_out


def predict(self, fh=None, X=None):
    """Forecast time series at future horizon.

    State required:
        Requires state to be "fitted".

    Accesses in self:
        Fitted model attributes ending in "_".
        self.cutoff, self._is_fitted

    Writes to self:
        Stores fh to self.fh if fh is passed and has not been passed previously.

    Parameters
    ----------
    fh : int, list, np.array or ForecastingHorizon, optional (default=None)
        The forecasting horizon encoding the time stamps to forecast at.
        if has not been passed in fit, must be passed, not optional
    X : time series in sktime compatible format, optional (default=None)
            Exogeneous time series to fit to
        Should be of same scitype (Series, Panel, or Hierarchical) as y in fit
        if self.get_tag("X-y-must-have-same-index"), X.index must contain fh.index
        there are no restrictions on number of columns (unlike for y)

    Returns
    -------
    y_pred : time series in sktime compatible data container format
        Point forecasts at fh, with same index as fh
        y_pred has same type as the y that has been passed most recently:
            Series, Panel, Hierarchical scitype, same format (see above)
    """
    # handle inputs

    self.check_is_fitted()
    fh = self._check_fh(fh)

    # input check and conversion for X
    X_inner = self._check_X(X=X)

    # we call the ordinary _predict if no looping/vectorization needed
    if not self._is_vectorized:
        y_pred = self._predict(fh=fh, X=X_inner)
    else:
        # otherwise we call the vectorized version of predict
        y_pred = self._vectorize("predict", X=X_inner, fh=fh)

    # convert to output mtype, identical with last y mtype seen
    y_out = convert_to(
        y_pred,
        self._y_metadata["mtype"],
        store=self._converter_store_y,
        store_behaviour="freeze",
    )
    y_out = self._validate_y_out(y_out)

    return y_out

```

**Comment by fkiraly** [comment]
Hm, I'd be against but not strongly, reasons:

* complicates base contract and extension template, adds a private extension point
* muddles layer separation between logic (output validation for modelling purpose) and interface (base class)
* does not cover the condition in the `TransformIf` case, that inherits from `BaseTransformer`

**Comment by ninedigits** [comment]
> Hm, I'd be against but not strongly, reasons:
> 
> * complicates base contract and extension template, adds a private extension point
> * muddles layer separation between logic (output validation for modelling purpose) and interface (base class)
> * does not cover the condition in the `TransformIf` case, that inherits from `BaseTransformer`

I agree, there's probably a better way. Do you have any more thoughts on how you'd implement what you're thinking, and what it might look like with an example?

**Comment by fkiraly** [comment]
> Do you have any more thoughts on how you'd implement what you're thinking, and what it might look like with an example?

Really just rough sketches. The "simplest" is moving the "condition" part out of `TransformIf`.

### Usage

Modified example of `TransformIf`

```python
seasonal = SeasonalityACF(candidate_sp=12) 
deseason = Deseasonalizer(sp=12) 
has_seasonality = Condition((seasonal, "sp", "!=", 1)
cond_deseason = TransformIf(has_seasonality, deseason) 
y_hat = cond_deseason.fit_transform(y)
```

Example for `FallbackForecaster`

```python
forecasters = [
    ("poly", PolynomialTrendForecaster()),
    ("naive", NaiveForecaster())
]

fail_condition1 = Condition(residuals, "mape", ">", 100)
fail_condition2 = Condition(relative_volatility, "var", ">", 10)

forecaster = FallbackForecaster(forecasters=forecasters, conditions=[fail_condition1, fail_condition2])
```

### possible docstring
```python
class Condition:
    """Condition parameter estimator.

    This compositor allows to specify a condition based on comparison.

    Parameters
    ----------
    if_estimator : sktime estimator, must have `fit`
        sktime estimator to fit and apply to series.
        this is a "blueprint" estimator, state does not change when `fit` is called
    param : str, optional, default = first boolean parameter of fitted if_estimator
    condition : str, optional, default = "bool"
        condition that defines whether self behaves like `then_est` or `else_est`
        this estimator behaves like `then_est` iff:
        "bool" = if `param` is True
        ">", ">=", "==", "<", "<=", "!=" = if `param condition condition_value`
    condition_value : required for some conditions, see above; otherwise optional
```

**Comment by ninedigits** [comment]
So it would follow the same structure as TransformIf? I need to better familiarize myself with that class.

**Comment by fkiraly** [comment]
> So it would follow the same structure as TransformIf?

That's an option - I split the class in two, i.e., condition and what to do given the condition.

**Comment by ninedigits** [comment]
So I think this is a great, but I'm not sure what the best way to approach it would be. I'm wondering if it's worth letting it sit for a bit?

**Comment by fkiraly** [comment]
hm, maybe.
Perhaps worth copy-pasting parts of `TransformIf` and turning it into `Condition` estimator with a simple boolean fitted param?

**Comment by ninedigits** [comment]
So I've been experimenting with `TransformIf`, and from what I understand, the `if` component functions by accessing fitted parameters. This implies it can be utilized with any entity that possesses this attribute, correct?

For instance, let's say I need to limit my forecasts with a `LogitTransformer`, particularly when forecasting a rate. There might be scenarios where the `BoxcoxTransformer` is best suited at lambda=0, leading to a double log scenario in the transformation process. To circumvent this, we could employ the following strategy with `TransformIf`:

```python3
TransformedTargetForecaster([
    ("logit", ScaledLogitTransformer(0, 1)),
    # Applies boxcox only if lambda not  equal to 0
    ("boxif", TransformIf(BoxCoxTransformer(), "lambda", "!=", 0, BoxCoxTransformer()))
    ("forecaster", ...)
])
```
If my understanding is accurate, and the fallback condition for `FallbackForecaster` would operate in a similar manner, then a specific set of parameters must be accessible for the `Condition` to evaluate whether to trigger or not. How do we know that any given estimator will have the necessary fitted parameter attributes to perform the check?

**Comment by fkiraly** [comment]
> This implies it can be utilized with any entity that possesses this attribute, correct?

yes, the attribute passed as the second argument.

> How do we know that any given estimator will have the necessary fitted parameter attributes to perform the check?

Can you make your question precise? Which estimator needs to have which parameter?

I was thinking that (a) `Condition` has a fitted parameter `"condition"`, and (b) an argument that can name the parameter of the underlying estimator, like `"lambda"`.

**Comment by ninedigits** [comment]
Revisiting this thread. Rather than try and explain my previous thoughts, let me take a step and lay things out the way I understand it. 

```python3
forecasters = [
    ("f1", PolynomialTrendForecaster()),
    ("f2", NaiveForecaster())
]

fail_condition1 = Condition(residuals, "mape", ">", 100)
fail_condition2 = Condition(relative_volatility, "var", ">", 10)

forecaster = FallbackForecaster(forecasters=forecasters, conditions=[fail_condition1, fail_condition2])

```

In the example you provided, we have two forecasters, `f1` and `f2`. In additional to the original design of `FallbackForecaster`, `f1` will fail if residuals have a mean absolute percent error of greater than 100, or if the variance is greater than 10. Same applies to `f2`, but `f1` is attempted first. 

My questions are around how the “condition” is calculated. 


Are the `residuals` and `relative_volatility` components calculated somewhere else as separate estimators? Example:

```python3
fail_condition1 = Condition(EvaluateResiduals(), "mape", ">", 100)
```

Would it make more sense to have something like, `FallbackIf` or `RaiseIf`? Really just a nomenclature thing here. 

```python3
fail_condition1 = FallbackIf(EvaluateResiduals(), "mape", ">", 100)
```

It’s possible I’m looking at this the completely wrong way.

**Comment by fkiraly** [comment]
> Are the `residuals` and `relative_volatility` components calculated somewhere else as separate estimators?

Yes, that's what I was imagining. Could be just a `Detrender` (which already computes residuals)

> Would it make more sense to have something like, `FallbackIf` or `RaiseIf`? Really just a nomenclature thing here.

I'd say, no - because the `Condition` is just an abstraction of a condition when sth shoud happen. That you want to move to the next forecaster in that case, is done by `FallbackForecaster`.

At least in my design, where the abstraction is based on noticing that one may want to do different things given a condition - eg.,. `TransformIf` or `FallbackForecaster` (which is a combination of "fallback if a condition, then this is the list of fallbacks")
